### ISSUE TITLE
fonts of third party css in dist/ root

### ISSUE THREAD

**Issue by maxkarkowski** [angular/angular-cli#3143]
node --version
v4.6.0
npm --version
3.10.8
angular-cli@1.0.0-beta.16

i have my own css library which i added in the angular-cli.json.

this library hast a font-face in its css like so:
```
@font-face {
    font-family: 'Interstate';
    src: url("**../fonts/**interstate-bold/86bef0b5-fa75-4ca3-8394-cb7b5a474a45-2.eot");
    font-style: normal;
    font-weight: bold
}
```

so i assumed if i run ng build the fonts would appear in a font folder or somehow in the assets folder.

but if i run ng build the font will placed in the dist/root folder like so

```
dist/
   fontfile.eot
```
in my case this looks like this
![dist folder](https://cloud.githubusercontent.com/assets/622841/20303272/f5c7a994-ab2a-11e6-8fe5-6b42ff77d98e.png)

can i somehow specify it that assets like this will not be placed in the root?

**Comment by Mage111** [comment]
I also have this problem in angular-cli@1.0.0-beta.16
with fonts that refer to assets folder
![image](https://cloud.githubusercontent.com/assets/23486790/20324831/ce33afbc-ab8a-11e6-9699-07b26e7527d7.png)

**Comment by Meligy** [comment]
Assets changed in a bit recently. Can you please try it out with beta.19-3 or beta.20-4?

**Comment by Mage111** [comment]
I upgraded to beta.20-4 and the problem still occurred.
All the fonts placed and referenced to assets folder.

[image: תמונה מוטבעת 1]

2016-11-16 22:08 GMT+02:00 Meligy, GuruStop.NET <notifications@github.com>:

**Comment by Meligy** [comment]
What's the `assets` property in your `angular-cli.json` like?

**Comment by Mage111** [comment]
Assets is folder in src folder. all resources referred to this folder.
angular-cli.json look like:

  "root": "src",
      "outDir": "dist",
      "assets": ["assets"],
      "index": "index.html",

**Comment by Meligy** [comment]
And you are sure you don't have copies of the files in `src/assets`?

**Comment by Mage111** [comment]
yes, I have placed the resources files in src/assets. in the build (prod), the folder src/assets has been copied to dist/assets. it works as expected for images, but not for fonts.

**Comment by Mage111** [comment]
I realize that something is wrong in my project structure. but what? Could you help me please?

**Comment by fellsquare** [comment]
I am facing the same issue. I generated the project using cli commands and didn't modify the structure. Is that an issue with cli itself or some specific configuration needs to be done to make it work?

**Comment by prafful-garg** [comment]
@maxkarkowski  can you please let me know if you have fixed this issue, as i am also facing the same issue and unable to fix.

**Comment by Mage111** [comment]
Not fixed yet.

**Comment by punekiran** [comment]
Any update on this, I'm facing the same issue.

**Comment by raoghuge** [comment]
Is this fixed yet. I am yet facing the same issue.

**Comment by mecheri** [comment]
I am also facing the same issue.

**Comment by filipesilva** [comment]
This is intended behaviour:
> So if you reference fonts in css, they'll just be processed/hashed/etc and put in the root.

**Comment by Starscream27** [comment]
@filipesilva Is there any way to change this behaviour ? My use case is I want to serve all assets (fonts referenced in the css included) from a CDN.

> This is intended behaviour:
> So if you reference fonts in css, they'll just be processed/hashed/etc and put in the root.

**Comment by smoke** [comment]
I have faced, the same problem and as far as I have looked there is no option to do so.
I ended up with a custom script doing that.
In short:
 - on ng build I set --deploy-url to '/__will-be-replaced-by-cdn-path__/'
 - after the build I replace '/__will-be-replaced-by-cdn-path__/' in all files with the real absolute URL to the cdn, this replaces references to images in the templates also.
 - after that you should place everything on the cdn, but the index.html file, you can serve it from wherever needed

I really hope this is supported from the cli any time soon.

**Comment by Starscream27** [comment]
@smoke Would you agree to share the code / script you used in order to accomplish this ?

**Comment by maximelafarie** [comment]
Always facing this issue in Angular 5. 
@filipesilva how's about the fix?

**Comment by chrisfoster78** [comment]
I really think this issue should to be reopened.  There is an option within the .angular-cli.json to output styles into a subfolder but this will break the font path as the fonts are always output to the root folder.

```
"styles": [
  {
    "input": "themes/theme1.scss",
    "output": "assets/theme1",
  }
]
```

I have tried:

```
"assets": [
  "assets",
  {
    "glob": "**/*", 
    "input": "../node_modules/@fortawesome/fontawesome-free-webfonts/webfonts", 
    "output": "/assets/" 
  }
],
```

but this fails with error:

```
An asset cannot be written to a location outside the project.
```

**Comment by johanroug** [comment]
It works for me now
The trick is to make absolute path in CSS
$fontpath: "/assets/fonts/";

```
@font-face {
	font-family: 'myFont';
	src: url($fontpath + myFont.eot');
	src: url($fontpath + myFont?#iefix') format('embedded-opentype'),
		url($fontpath + 'myFont.woff2') format('woff2'),
		url($fontpath + 'myFont.woff') format('woff');
	font-weight: normal;
	font-style: normal;
}
```

and in Angular.json file

```
"assets": [
              "src/assets/fonts"
            ],
```

**Comment by ldavidson** [comment]
I was having an issue with this and have found the solution for me to this issue.
Looked at the angular starter webpack config files I figured out how to change the output folder. This will not require any changes to embedded fonts.
````javascript
{
  test: /\.(eot|woff2?|svg|ttf)([\?]?.*)$/,
  use: {
    loader: 'file-loader',
    options: {
      outputPath: '<the folder you want your fonts to go>'
    }
  }
}
````

**Comment by mauro1998** [comment]
@johanroug is right, this worked for me as well;

`_fonts.scss`:

    $font-path: '/assets/fonts/';

`angular-cli.json`:

    "assets": [
        "assets",
        "favicon.ico"
     ],

**Comment by sebastian-zarzycki-es** [comment]
Most solutions people paste here are solutions for their local files, not for css fonts imported from node_modules. Lots of confusion. For local files, just use absolute paths. For node_modules - I don't know of any.

@ldavidson , I've tried that and it didn't seem to influence anything (Angular 7). It ignored anything I've put into outputPath.

**Comment by TheILPlace** [comment]
> Most solutions people paste here are solutions for their local files, not for css fonts imported from node_modules. Lots of confusion. For local files, just use absolute paths. For node_modules - I don't know of any.
> 
> @ldavidson , I've tried that and it didn't seem to influence anything (Angular 7). It ignored anything I've put into outputPath.

absolute path is not a good option, if you use ng build with --base-href , as it will put the basehref folder inside the css files. which will cause problems if you deploy to multiple locations under different folders

**Comment by LeTranAnhVu** [comment]
@sebastian-zarzycki-es  : Hi , I'm facing the same issue. I have solved the local files by using absolute path, but I stuck in some plugin or lib in the bower_components. 
Have you found the solutions?

**Comment by sebastian-zarzycki-es** [comment]
Nope. I would expect CLI handles it properly on its own. Ideally sooner, than later.

**Comment by naman1901** [comment]
This should be re-opened and fonts moved to a non-root folder in dist. Otherwise, it just looks super-unorganized.

**Comment by RscAnonymous** [comment]
I am using a third party library in my angular project, which might be using absolute path thus my ng prod is placing all the fonts and images used by it in the root directory. Any workaround on this?

**Comment by jcimoch** [comment]
I think this should be reopened

**Comment by nilesh1chavan** [comment]
Please reopen this and if this any solution, put it here
